ambry-api/src/main/java/com.github.ambry/account/Account.java

```diff
@@ -0,0 +1,312 @@
+ *   Below lists all the metadata versions and their formats:
+ * </p>
+ * <pre><code>
+ * Version 1:
+ * {
+ *   "accountId": 101,
+ *   "accountName": "MyAccount",
+ *   "containers": [
+ *     {
+ *       "containerName": "MyPrivateContainer",
+ *       "description": "This is my private container",
+ *       "isPrivate": "true",
+ *       "containerId": 0,
+ *       "version": 1,
+ *       "status": "ACTIVE",
+ *       "parentAccountId": "101"
+ *     },
+ *     {
+ *       "containerName": "MyPublicContainer",
+ *       "description": "This is my public container",
+ *       "isPrivate": "false",
+ *       "containerId": 1,
+ *       "version": 1,
+ *       "status": "ACTIVE",
+ *       "parentAccountId": "101"
+ *     }
+ *   ],
+ *   "version": 1,
+ *   "status": "ACTIVE"
+ * }
+ * </code></pre>
+ * <p>
+ *   An account object is immutable. To update an account, refer to {@link AccountBuilder} for how to build a new
+ *   account object with updated field(s).
+ * </p>
+ */
+public class Account {
+  // static variables
+  static final String ACCOUNT_METADATA_VERSION_KEY = "version";
+  static final String ACCOUNT_ID_KEY = "accountId";
+  static final String ACCOUNT_NAME_KEY = "accountName";
+  static final String ACCOUNT_STATUS_KEY = "status";
+  static final String CONTAINERS_KEY = "containers";
+  static final short ACCOUNT_METADATA_VERSION_1 = 1;
+  static final short HIGHEST_ACCOUNT_METADATA_VERSION = ACCOUNT_METADATA_VERSION_1;
+  // account member variables
+  private final Short id;
+  private final String name;
+  private AccountStatus status;
+  // internal data structure
+  private final Map<Short, Container> containerIdToContainerMap = new HashMap<>();
+  private final Map<String, Container> containerNameToContainerMap = new HashMap<>();
+
+  /**
+   * Constructing an {@link Account} object from account metadata.
+   * @param metadata The metadata of the account in JSON.
+   * @throws JSONException If fails to parse metadata.
+   */
+  Account(JSONObject metadata) throws JSONException {
+    if (metadata == null) {
+      throw new IllegalArgumentException("metadata cannot be null.");
+    }
+    short metadataVersion = (short) metadata.getInt(ACCOUNT_METADATA_VERSION_KEY);
+    switch (metadataVersion) {
+      case ACCOUNT_METADATA_VERSION_1:
+        id = (short) metadata.getInt(ACCOUNT_ID_KEY);
+        name = metadata.getString(ACCOUNT_NAME_KEY);
+        status = AccountStatus.valueOf(metadata.getString(ACCOUNT_STATUS_KEY));
+        checkRequiredFieldsForBuild();
+        JSONArray containerArray = metadata.getJSONArray(CONTAINERS_KEY);
+        Collection<Container> containers = new ArrayList<>();
+        if (containerArray != null) {
+          for (int index = 0; index < containerArray.length(); index++) {
+            JSONObject containerMetadata = containerArray.getJSONObject(index);
+            Container container = new Container(containerMetadata);
+            containers.add(container);
```

Comment:
I got your point to do the checks inside of the for loop. Will come to you and discuss if this would be a better place. I made some attempt and there is pros and cons in both way.